### PR TITLE
[CH-279] Support more window functions, lead/lag/dense_rank

### PR DESCRIPTION
**Pull request by lgbo-ustc**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)
Support more window functions, lead/lag/dense_rank

`first_value` and `last_value` have different effects in CH, will be added later

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
unit tests

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

test by CH[[322]]

### CODE REVIEW

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala [hunk 1]
```diff
@@ -44,7 +47,27 @@ object VeloxBackendSettings extends BackendSettings {
...
+      windowFunctions.foreach(
+        func => {
+          val aliasExpr = func.asInstanceOf[Alias]
+          val wExpression = aliasExpr.child.asInstanceOf[WindowExpression]
+          wExpression match {
+            case rowNumber: RowNumber => allSupported = allSupported & true
+            case aggExpression: AggregateExpression => allSupported = allSupported & true
+            case rank: Rank => allSupported = allSupported & true
+            case lead: Lead => allSupported = allSupported & true
+            case lag: Lag => allSupported = allSupported & true
```
Comment (PHILO-HE) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:62]: Currently, for velox backend, Lead/Lag are not supported. We should set `allSupported` to false here, right?

Comment (lgbo-ustc) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:62]: Yes

**Comment by PHILO-HE** [comment]
For velox backend UT failure, `SPARK-14664: Decimal sum/avg over window should work`, I found it was caused by instance cast failure. It seems Spark optimized the plan with `MakeDecimal` or `cast` used as the parent plan of the window expression. 

The below changes can fix this issue (may need some refinement). CHBackend.scala should also do some similar changes (this UT is excluded by CH, so the failure was not exposed in CH backend).

```
diff --git a/backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala b/backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala
index 6f9d6c29..504405c2 100644
--- a/backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala
+++ b/backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala
@@ -20,7 +20,7 @@ package io.glutenproject.backendsapi.velox
 import io.glutenproject.GlutenConfig
 import io.glutenproject.backendsapi._
 import org.apache.spark.sql.catalyst.plans.JoinType
-import org.apache.spark.sql.catalyst.expressions.{Alias, NamedExpression, Rank, RowNumber, WindowExpression}
+import org.apache.spark.sql.catalyst.expressions.{Alias, Expression, MakeDecimal, NamedExpression, Rank, RowNumber, WindowExpression}
 import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression
 import org.apache.spark.sql.execution.datasources.FileFormat
 import org.apache.spark.sql.execution.datasources.orc.OrcFileFormat
@@ -48,14 +48,25 @@ object VeloxBackendSettings extends BackendSettings {
   override def supportExpandExec(): Boolean = true
   override def needProjectExpandOutput: Boolean = true
   override def supportSortExec(): Boolean = true
+
+  def getWindowExpression(expr: Expression, windowList: java.util.List[WindowExpression]): Unit = {
+      expr match {
+        case w: WindowExpression => windowList.add(w)
+        case other => other.children.foreach(child => getWindowExpression(child, windowList))
+      }
+  }
+
   override def supportWindowExec(windowFunctions: Seq[NamedExpression]): Boolean = {
     var allSupported = true
     breakable {
       windowFunctions.foreach(
         func => {
           val aliasExpr = func.asInstanceOf[Alias]
-          val wExpression = aliasExpr.child.asInstanceOf[WindowExpression]
-          wExpression match {
+          val windowList = new java.util.LinkedList[WindowExpression]()
+          getWindowExpression(aliasExpr.child, windowList)
+          // Only one window expression is expected?
+          val wExpression = windowList.get(0)
+            wExpression match {
             case rowNumber: RowNumber => allSupported = allSupported & true
             case aggExpression: AggregateExpression => allSupported = allSupported & true
             case rank: Rank => allSupported = allSupported & true
```

**Comment by lgbo-ustc** [comment]
> supportWindowExec

@PHILO-HE  Is this fix merged ?

**Comment by PHILO-HE** [comment]
> > supportWindowExec
> 
> @PHILO-HE Is this fix merged ? If not , I will refine in this pr

Not merged. Please directly do some refinement in this PR. Thanks!

**Comment by lgbo-ustc** [comment]
> Run cd /home/qwe1398775315/code/ClickHouse
error: The following untracked working tree files would be overwritten by checkout:
	contrib/libdivide/CMakeLists.txt
	contrib/libdivide/LICENSE.txt
	contrib/libdivide/libdivide.h
Please move or remove them before you switch branches.

😓

**Review by zzcclp — APPROVED** [review]
LGTM

**Pull request merged**